### DISCUSSION TITLE
Server refuses to start with Steam error on odin install

### DISCUSSION THREAD

**Discussion by alexmaras** [mbround18/valheim-docker#456] [Help!]
## System Information

Operating System: Linux - OpenSuse MicroOS - https://microos.opensuse.org/
Docker Version: Docker version 20.10.6-ce, build 8728dd246c3a
Docker Compose Version: docker-compose version 1.28.5, build unknown
User & Group On Host: 1000:1000
Are you running with BepInEx? No

## Issue Description

Write up what is the issue you are experiencing. 
`odin install` when bringing up the container fails

Logs from running `docker-compose up`:
```
alex@atlas:~/docker/valheim> docker-compose up
Creating network "valheim_default" with the default driver
Creating valheim_valheim_1 ... done
Attaching to valheim_valheim_1
valheim_1  | [Valheim][root] : ###########################################################################
valheim_1  | [Valheim][root] : Valheim Server - Tue Sep 28 13:32:52  2021
valheim_1  | [Valheim][root] : Initializing your container...
valheim_1  | [Valheim][root] : You must be in development. Good luck!
valheim_1  | [Valheim][root] : ###########################################################################
valheim_1  | [Valheim][root] : Switching UID and GID
valheim_1  | usermod: no changes
valheim_1  | [Valheim][root] : 
valheim_1  | [Valheim][root] : 
valheim_1  | [Valheim][root] : Auto Update Enabled...
valheim_1  | [Valheim][root] : Auto Update Schedule: "0 3 * * *"
valheim_1  | [Valheim][root] : Auto Backup Enabled...
valheim_1  | [Valheim][root] : Auto Backup Schedule: "*/15 * * * *"
valheim_1  | [Valheim][root] : Setting up file systems
valheim_1  | [Valheim][root] : Navigating to steam home...
valheim_1  | [Valheim][root] : Launching as steam...
valheim_1  | [Valheim][steam]: ###########################################################################
valheim_1  | [Valheim][steam]: Valheim Server - Tue Sep 28 13:32:53  2021
valheim_1  | [Valheim][steam]: STEAM_UID 1000 - STEAM_GUID 1000
valheim_1  | [Valheim][steam]: Installing Valheim via odin 1.4.2...
valheim_1  | [Valheim][steam]: ###########################################################################
valheim_1  | [Valheim][steam]: Variables loaded.....
valheim_1  | [Valheim][steam]: Port: 2456
valheim_1  | [Valheim][steam]: Name: Created With Valheim Docker
valheim_1  | [Valheim][steam]: World: Dedicated
valheim_1  | [Valheim][steam]: Public: 0
valheim_1  | [Valheim][steam]: Password: (REDACTED)
valheim_1  | [Valheim][steam]: Running Install...
valheim_1  | [ODIN][INFO]  - Installing 896660 to /home/steam/valheim
valheim_1  | [ODIN][ERROR] - Failed to find steamcmd in path
valheim_1  | [ODIN][ERROR] - Checking for script under steam user.
valheim_1  | [ODIN][INFO]  - Executing: /home/steam/steamcmd/steamcmd.sh .....
valheim_1  | [ODIN][INFO]  - Using steamcmd script at /home/steam/steamcmd/steamcmd.sh
valheim_1  | Redirecting stderr to '/home/steam/Steam/logs/stderr.txt'
valheim_1  | [  0%] Checking for available updates...
valheim_1  | [----] Downloading update (51 of 54,020 KB)...
valheim_1  | [  0%] Downloading update (853 of 54,020 KB)...
valheim_1  | [  1%] Downloading update (2,949 of 54,020 KB)...
valheim_1  | [  5%] Downloading update (4,827 of 54,020 KB)...
valheim_1  | [  8%] Downloading update (6,628 of 54,020 KB)...
valheim_1  | [ 12%] Downloading update (8,501 of 54,020 KB)...
valheim_1  | [ 15%] Downloading update (10,696 of 54,020 KB)...
valheim_1  | [ 19%] Downloading update (12,562 of 54,020 KB)...
valheim_1  | [ 23%] Downloading update (14,413 of 54,020 KB)...
valheim_1  | [ 26%] Downloading update (16,284 of 54,020 KB)...
valheim_1  | [ 30%] Downloading update (18,291 of 54,020 KB)...
valheim_1  | [ 33%] Downloading update (19,989 of 54,020 KB)...
valheim_1  | [ 37%] Downloading update (21,783 of 54,020 KB)...
valheim_1  | [ 40%] Downloading update (23,627 of 54,020 KB)...
valheim_1  | [ 43%] Downloading update (25,682 of 54,020 KB)...
valheim_1  | [ 47%] Downloading update (27,554 of 54,020 KB)...
valheim_1  | [ 51%] Downloading update (28,976 of 54,020 KB)...
valheim_1  | [ 53%] Downloading update (30,430 of 54,020 KB)...
valheim_1  | [ 56%] Downloading update (32,386 of 54,020 KB)...
valheim_1  | [ 59%] Downloading update (34,249 of 54,020 KB)...
valheim_1  | [ 63%] Downloading update (35,820 of 54,020 KB)...
valheim_1  | [ 66%] Downloading update (36,632 of 54,020 KB)...
valheim_1  | [ 67%] Downloading update (37,699 of 54,020 KB)...
valheim_1  | [ 69%] Downloading update (38,841 of 54,020 KB)...
valheim_1  | [ 71%] Downloading update (39,884 of 54,020 KB)...
valheim_1  | [ 73%] Downloading update (40,883 of 54,020 KB)...
valheim_1  | [ 75%] Downloading update (42,019 of 54,020 KB)...
valheim_1  | [ 77%] Downloading update (43,086 of 54,020 KB)...
valheim_1  | [ 79%] Downloading update (44,169 of 54,020 KB)...
valheim_1  | [ 81%] Downloading update (45,256 of 54,020 KB)...
valheim_1  | [ 83%] Downloading update (46,556 of 54,020 KB)...
valheim_1  | [ 86%] Downloading update (47,762 of 54,020 KB)...
valheim_1  | [ 88%] Downloading update (48,947 of 54,020 KB)...
valheim_1  | [ 90%] Downloading update (50,283 of 54,020 KB)...
valheim_1  | [ 93%] Downloading update (51,456 of 54,020 KB)...
valheim_1  | [ 95%] Downloading update (52,561 of 54,020 KB)...
valheim_1  | [ 97%] Downloading update (53,742 of 54,020 KB)...
valheim_1  | [ 99%] Downloading update (54,020 of 54,020 KB)...
valheim_1  | [100%] Download complete.
valheim_1  | [----] Installing update...
valheim_1  | [----] Extracting package...
valheim_1  | [----] Extracting package...
valheim_1  | [----] Extracting package...
valheim_1  | [----] Extracting package...
valheim_1  | [----] Installing update...
valheim_1  | [----] Installing update...
valheim_1  | [----] Installing update...
valheim_1  | [----] Installing update...
valheim_1  | [----] Installing update...
valheim_1  | [----] Installing update...
valheim_1  | [----] Installing update...
valheim_1  | [----] Installing update...
valheim_1  | [----] Cleaning up...
valheim_1  | [----] Update complete, launching Steamcmd...
valheim_1  | Redirecting stderr to '/home/steam/Steam/logs/stderr.txt'
valheim_1  | [  0%] Checking for available updates...
valheim_1  | [----] Verifying installation...
valheim_1  | Steam Console Client (c) Valve Corporation
valheim_1  | -- type 'quit' to exit --
valheim_1  | Loading Steam API...src/common/steam/client_api.cpp (579) : Assertion Failed: ClientAPI_InitGlobalInstance: InternalAPI_Init_Internal failed, most likely because you are missing a 32-bit dependency of steamclient.so (the Steam client is a 32-bit app).
valheim_1  | 
valheim_1  | FAILED (1)
valheim_1  | [ODIN][INFO]  - Exited with http code: 3
valheim_1  | cp: cannot create regular file '/home/steam/valheim/linux64/': Not a directory
valheim_1  | [Valheim][steam]: Initializing Variables....
valheim_1  | [ODIN][ERROR] - Failed to find server executable! Please run `odin install`
alex@atlas:~/docker/valheim> 
```

docker-compose.yml file:
```
version: "3"
services:
  valheim:
    restart: unless-stopped
    image: mbround18/valheim:latest
    ports:
      - 2456:2456/udp
      - 2457:2457/udp
      - 2458:2458/udp
    environment:
      - PORT=2456
      - NAME="Created With Valheim Docker"
      - WORLD="Dedicated"
      - PASSWORD="*********"
      - TZ="Australia/Perth"
      - PUBLIC=0
      - AUTO_UPDATE=1
      - AUTO_UPDATE_SCHEDULE="0 3 * * *"
      - AUTO_BACKUP=1
      - AUTO_BACKUP_SCHEDULE="*/15 * * * *"
      - AUTO_BACKUP_DAYS_TO_LIVE=3
      - AUTO_BACKUP_ON_UPDATE=1
      - AUTO_BACKUP_ON_SHUTDOWN=1
      - UPDATE_ON_STARTUP=1
    volumes:
    - ./saves:/home/steam/.config/unity3d/IronGate/Valheim
    - ./server:/home/steam/valheim
    - ./backups:/home/steam/backups
```

I've deleted the `server` directory to clear things out in case something's corrupted - but nothing appears to work. I've run a `docker-compose pull` to ensure I'm on the latest version, and also tried rolling back to the previous version - same outcome

Any thoughts? The exact same `docker-compose.yml` with the same save is working on my main PC.

**Comment by mbround18** [comment]
@alexmaras does the path `/home/steam/valheim/valheim_server.x86_64` exist in the container?

**Reply by alexmaras** [reply]
No, only `logs` and `steamclient.so`. I assume it's what would be downloaded by `steamcmd` when starting the container normally, which is failing in the step where it shows
```
Loading Steam API...src/common/steam/client_api.cpp (579) : Assertion Failed: ClientAPI_InitGlobalInstance: InternalAPI_Init_Internal failed, most likely because you are missing a 32-bit dependency of steamclient.so (the Steam client is a 32-bit app).
```
Manually running `odin install` (using `docker-compose run valheim odin install`) also doesn't work - it results in the same error, and just stops at 
```
FAILED (1)
[ODIN][INFO]  - Exited with http code: 3
```
instead of then trying to start the server.

I'm wondering if this issue is rooted in the base dependency (`cm2network/steamcmd`)?

**Reply by alexmaras** [reply]
Also, doing `docker-compose run -u root --entrypoint bash valheim` to enter the container, then `su steam` and running `/home/steam/steamcmd/steamcmd.sh` results in exactly the same error, so it's definitely an issue with `steamcmd` in the container

**Reply by mbround18** [reply]
what tag is your container running on? does the container have adequate network access? I ask because I use this image to run ~20 valheim servers atm and I am not seeing an issue with steamcmd inside the container. Have you tried blowing away the contents of `/home/steam/valheim`?

**Reply by alexmaras** [reply]
I'm using `:latest`, and I've checked the digest against the latest image in dockerhub to ensure I've got the right one running. I just saw that a new `:latest` was pushed yesterday, I'm just updated to that now to ensure it's still broken for me, and it is. Digest is `db040388fed2c126ddb6f0f0b5f16a065414e0ffcf47ce298feb55dfec73d6ca`. 

Network access should be good - I can run `curl` in the container and get results, and the failure seems to be more in the `steamcmd` side than the valheim side - running `steamcmd.sh` manually results in the steam update downloading completely, then the command failing to start because of what it states as missing 32-bit dependencies. It doesn't list which dependencies annoyingly. 

I certainly haven't had an issue with the image on my local PC either, so this is baffling me a bit. I'm using `ldd` to try to track down any missing dependencies at the moment.

**Reply by alexmaras** [reply]
I've just run the base image (`docker run -it cm2network/steamcmd:root bash`) and have got the same error. I'm going to go over to that repo and ask the question there instead, as this seems to be unrelated to the changes you've made on top of that Dockerfile.

Here's a link to the issue - I've found that manually updating `curl` to the latest version in that dockerfile has it enter the `Steam>` prompt as expected:

https://github.com/CM2Walki/steamcmd/issues/28

**Reply by alexmaras** [reply]
Can confirm - manually building that image (fixing up the curl version as mentioned in the comment on that issue) and then using your repo with the dockerfile edited to be `FROM locally-built-steamcmd:latest` works as expected - server is running, and the save functions as normal with the updates valheim server. 

Weird that `curl` worked fine in the container when running, but that I had to bump the version to get `steamcmd.sh` to work.

**Reply by alexmaras** [reply]
If you end up running into the same issue down the road, I would suggest that this image could provide a replacement base: https://hub.docker.com/r/steamcmd/steamcmd
It seems to get updated frequently. If you're interested, I'd happy make up a PR to get that work done to swap over to that image's `ubuntu-20` tag. It shouldn't be a lot of effort to get that working - just slightly different steamcmd paths.

**Reply by mbround18** [reply]
@alexmaras would you mind giving the following image a try? it uses the steamcmd image you recommeded.
```
mbround18/valheim:dev
```

https://hub.docker.com/layers/mbround18/valheim/dev/images/sha256-bbc1090ed8576aaad24fa7a160715dc38372b6da2973f40c61802c9a6475369f?context=repo

**Reply by alexmaras** [reply]
![image](https://user-images.githubusercontent.com/7509232/135697251-d5e37b78-2da6-4178-a484-0e0cd0946f48.png)

I had a minor error, not something particular problematic - it's just the steamcmd dumps. I suppose it just means it wouldn't be able to create dumps to send to valve. 

Everything else worked perfectly, and it came up first time with my existing save. 

Did you end up using the alpine image like you suggested, or did you have issues? I find alpine to be a solid container base (I use it myself if I can) but I figured for a direct swap being able to use `apt` would reduce the chance of issues

**Reply by mbround18** [reply]
Hey @alexmaras apologies for the delay, hectic weekend, I wanted to switch to alpine but i kept getting a weird error whenever i tried to compile and copy over files. 

```
standard_init_linux.go:228: exec user process caused: exec format error
```

and it didnt seem to copy files over appropriately

**Reply by alexmaras** [reply]
mmm, seems like that's either an architecture issue (unlikely - there's no ARM in the mix anywhere) or something where a shebang isn't quite right. Bit odd. The ubuntu one is working great for me at the moment. I might spend a little time in the week mucking about with an alpine one if I get some time. If you have a branch with the attempted alpine version and don't mind pushing it up, I'd be happy to do some troubleshooting.

**Reply by samtietze** [reply]
Just spotted this in the Discussions, and it's quite similar for me. I made a post in a different thread: https://github.com/mbround18/valheim-docker/discussions/468#discussioncomment-1528764
